pennylane/gradients/spsa_gradient.py

```diff
@@ -0,0 +1,573 @@
+
+
+        Devices that have a shot vector defined can also be used for execution, provided
+        the ``shots`` argument was passed to the transform:
+
+        >>> shots = (10, 100, 1000)
+        >>> dev = qml.device("default.qubit", wires=2, shots=shots)
+        >>> @qml.qnode(dev)
+        ... def circuit(params):
+        ...     qml.RX(params[0], wires=0)
+        ...     qml.RY(params[1], wires=0)
+        ...     qml.RX(params[2], wires=0)
+        ...     return qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(0))
+        >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
+        >>> qml.gradients.spsa_grad(circuit, shots=shots, h=1e-2)(params)
+        (((array(0.), array(0.), array(0.)), (array(0.), array(0.), array(0.))),
+         ((array(-1.4), array(-1.4), array(-1.4)),
+          (array(2.548), array(2.548), array(2.548))),
+         ((array(-1.06), array(-1.06), array(-1.06)),
+          (array(1.90588), array(1.90588), array(1.90588))))
+
+        The outermost tuple contains results corresponding to each element of the shot vector,
+        as is also visible by the increasing precision.
+        Note that the stochastic approximation and the fluctuations from the shot noise
+        of the device accumulate, leading to a very coarse-grained estimate for the gradient.
+    """
+    if argnum is None and not tape.trainable_params:
+        return _no_trainable_grad_new(tape, shots)
+
+    if validate_params:
+        if "grad_method" not in tape._par_info[0]:
+            gradient_analysis(tape, grad_fn=_spsa_grad_new)
+        diff_methods = grad_method_validation("numeric", tape)
+    else:
+        diff_methods = ["F" for i in tape.trainable_params]
+
+    if all(g == "0" for g in diff_methods):
+        return _all_zero_grad_new(tape, shots)
+
+    gradient_tapes = []
+    extract_r0 = False
+
+    coeffs, shifts = finite_diff_coeffs(n=n, approx_order=approx_order, strategy=strategy)
+
+    if 0 in shifts:
+        # Finite difference formula includes a term with zero shift.
+
+        if f0 is None:
+            # Ensure that the unshifted tape is appended to the gradient tapes
+            gradient_tapes.append(tape)
+            extract_r0 = True
+
+        # Skip the unshifted tape
+        shifts = shifts[1:]
+
+    method_map = choose_grad_methods(diff_methods, argnum)
+
+    indices = [
+        i for i, _ in enumerate(tape.trainable_params) if (i in method_map and method_map[i] != "0")
+    ]
+
+    tapes_per_grad = len(shifts)
+    all_coeffs = []
+    for idx_rep in range(num_directions):
+        direction = sampler(indices, len(tape.trainable_params), idx_rep)
+        inv_direction = qml.math.divide(
+            1, direction, where=(direction != 0), out=qml.math.zeros_like(direction)
+        )
+        # Use only the non-zero part of `direction` for the shifts, to skip redundant zero shifts
+        _shifts = qml.math.tensordot(h * shifts, direction[indices], axes=0)
```

Comment:
So here we select the directions that have trainable parameters?

Reply:
They are selected via `indices`, which is computed from the output of `choose_grad_methods` further up :) 

A bit of context (maybe this is clear already):
The way the `sampler` is supposed to work is that it obtains 
1. the indices within the list of trainable tape parameters that is supposed to be treated (e.g. trainable tape parameters that do not influence the output value of the QNode will be skipped)
2. the number of trainable tape parameters, including those that are not selected by `indices` in 1.
3. An iteration variable, which until now is mostly used in the dummy sampler in the test suite. `_rademacher_sampler` for example does not use this input at all.

and that it returns a `direction` vector that has the full size for all trainable tape parameters, but only the ones chosen by `indices` are actually filled with non-zero values.

In the tagged line itself we compute `_shifts`, which actually does _not_ have the full size of all trainable tape parameters, but just the same size as `indices`. This makes sure that the output Jacobian actually has the correct shape instead of returning zeros for the parameters that are not supposed to be differentiated.

Reply:
Great explanation thank you!

Reply:
@timmysilv Useful details here.